src/crimson/osd/object_context.cc

```diff
@@ -46,7 +46,8 @@ void ObjectContextRegistry::handle_conf_change(
 
 std::optional<hobject_t> resolve_oid(
   const SnapSet &ss,
-  const hobject_t &oid)
+  const hobject_t &oid,
+  bool check_clone_snaps)
```

Comment:
We can be more specific here.  For the trim case, we aren't mapping an arbitrary snapshot onto a set of clones as we are in the normal IO pathway.  We want the *exact* clone specified in the oid param.  Really, none of the code in resolve_oid is correct for that use case -- the std::lower_bound should be a std::find, the initial if branch shouldn't be possible, and snaps (which you have conditioned out) is incorrect.  Rather than modifying resolve_oid, which would create confusingly divergent behavior deep in the call chain, let's introduce a new method resolve_clone_direct() and call it from a new with_direct_clone_obc() method.  The price is some minor structural duplication of the locking machinery in with_head_and_clone_obc, but the semantics are otherwise genuinely independent.

Reply:
That is, in SnapTrimEvent::with_pg, we pull the coid out of snap_map.get_next_objects_to_trim, which I believe returns the exact hobject_t for the clone in question.  I might be missing something though.

Reply:
In fact, since I don't think the oid actually needs to be "resolved", we probably just want an ObjectContextLoader::with_clone_obc_direct() method which simply loads the clone obc directly based on the passed hobject_t (along with the head since IIRC there's an invariant that we always have the head loaded when a clone is loaded).

Reply:
@athanatos All done:-) Since `ObjectContextLoader::with_head_and_clone_obc()` is only called by SnapTrimEvent, I directly renamed it to `ObjectContectLoader::with_clone_obc_direct()` and removed the `resolve_oid()` operation in it. Please take a look again.